selftest/map-batch/main.go

```diff
@@ -0,0 +1,128 @@
+package main
+
+import "C"
+
+import (
+	"os"
+	"unsafe"
+
+	"encoding/binary"
+	"fmt"
+
+	bpf "github.com/aquasecurity/libbpfgo"
+)
+
+func resizeMap(module *bpf.Module, name string, size uint32) error {
+	m, err := module.GetMap("events")
+	if err != nil {
+		return err
+	}
+
+	if err = m.Resize(size); err != nil {
+		return err
+	}
+
+	if actual := m.GetMaxEntries(); actual != size {
+		return fmt.Errorf("map resize failed, expected %v, actual %v", size, actual)
+	}
+
+	return nil
+}
+
+func main() {
+
+	bpfModule, err := bpf.NewModuleFromFile("main.bpf.o")
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(-1)
+	}
+	defer bpfModule.Close()
+
+	if err = resizeMap(bpfModule, "events", 8192); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(-1)
+	}
+
+	bpfModule.BPFLoadObject()
+
+	testerMap, err := bpfModule.GetMap("tester")
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(-1)
+	}
+
+	keys := []uint32{1, 2, 3, 4, 5, 6, 8}
+	values := []uint32{2, 3, 4, 5, 6, 7, 9}
+
+	// Test batch update.
+	if err := testerMap.UpdateBatch(unsafe.Pointer(&keys[0]), unsafe.Pointer(&values[0]), uint32(len(keys))); err != nil {
+		fmt.Fprintf(os.Stderr, "testerMap.UpdateBatch failed: %v", err)
+		os.Exit(-1)
+	}
+	val, err := testerMap.GetValue(unsafe.Pointer(&keys[0]))
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "testerMap.GetValue failed: %v", err)
+		os.Exit(-1)
+	}
+	if binary.LittleEndian.Uint32(val) != values[0] {
+		fmt.Fprintf(os.Stderr, "testerMap.GetValue returned %v, expected %v", val, values[0])
+		os.Exit(-1)
+	}
+
+	// Test batch lookup in steps.
+	batchKeys := make([]uint32, 3)
+	nextKey := uint32(0)
+	prevKey := unsafe.Pointer(nil)
+	step := len(batchKeys)
+	for i := 0; i < 2; i++ {
+		if i > 0 {
+			// We're on step 2, so test the batch lookup by specifying the
+			// key to start from.
+			prevKey = unsafe.Pointer(&nextKey)
+		}
```

Comment:
Can you explain more of what's going on here? Why is there two steps? I see the documentation for BPF_MAP_LOOKUP_BATCH in uapi/linux/bpf.h but I don't fully understand why it's designed this way.

Reply:
Yeah, so the API is such that you can iterate over all the keys/values in batches. Essentially you provide a pointer to the last key seen, the next key to start with, and a buffer for keys/values.

That way if you have a map with max size of 1000 for example you can iterate over the entire thing in chunks instead of loading everything into memory. It's kind of a confusing API to be honest it took me a bit to wrap my head around it when implementing.

Reply:
Got it - that took some time to understand. Definitely worth documenting upstream in libbpf.

Reply:
After that this is good to go :+1:

Reply:
> Got it - that took some time to understand. Definitely worth documenting upstream in libbpf.

Yes! I literally had to read through the test code to understand how to use the API.

Reply:
Ok, let me know if the documentation I added is clear enough.